Add Snapshot testing

DawoodShahat opened the pull request:
This PR fixes #117 

### Pull Request checklist

<!-- Before submitting a PR for review, please address each item -->

- [x] The pull request has a descriptive title (and a reference to an issue it
      fixes, if applicable)
- [x] All tests and linter checks are passing
- [x] The pull request is free of merge conflicts

<!-- For more information, see CONTRIBUTING.md at the root of this repository -->

wlach commented on src/components/FilterInput.svelte:
```diff
@@ -1,5 +1,5 @@
 <script>
-  import throttle from "just-throttle";
+  const throttle = require("just-throttle");
```
I think we should probably not make these changes, as the `import` syntax is more intuitive to people. I gather you're seeing some kind of error here? Can you post a trace and/or Google the results? I'm guessing that we probably just need to add a pre/post processing step somewhere in the build configuration.

DawoodShahat replied:
You are right. Currently I'm getting an error similar to this issue [https://github.com/facebook/jest/issues/6229](https://github.com/facebook/jest/issues/6229). 

- This Issue arises when an external library is used e.g. `marked` or `just-throttle` that is not transpiled to make ES6 imports work with nodejs. 

- `babel-jest` ignores node_modules altogether. I tried different plugins to make it work e.g.  `@babel/plugin-transform-modules-commonjs`. I also tried different solutions provided in the aforementioned issue, but still facing the same issue.

- Here is what I get when I try to use `marked`
![Screenshot 2020-11-18 at 10 27 21 PM](https://user-images.githubusercontent.com/19612755/99565357-57aee600-29ed-11eb-8af7-440e4d7b2f6f.png)

wlach replied:
Ah ok. Went down a bit of a rabbit hole with this. After looking at things a bit more, I think this has to do with the particulars of how these two libraries work as you describe. In the case of `just-throttle`, the solution is probably just to use [lodash's throttle function](https://lodash.com/docs/4.17.15#throttle) instead (this will require depending on lodash, which we want to do anyway IMO). In the case of marked, it's actually just a minor change we need to make in how it's imported. I'll suggest it.

wlach replied:
Oops! I spoke too soon. We actually *can't* use require, as it's not recognized by client-side JavaScript:

<img width="520" alt="image" src="https://user-images.githubusercontent.com/20569/99595182-4a303680-29c2-11eb-8bcc-d25913c73ef1.png">

Do you want to just go ahead and add lodash to the repo as part of this PR? Just run `npm install lodash` and you should be good to go.

DawoodShahat replied:
Yes, I will make the suggested changes in the next commit. Thanks.

wlach reviewed (changes requested):
This looks great @DawoodShahat! I have some concerns about modifying the svelte files, but other than that this is looking very close.

Checking the CI logs, I noticed this warning:


```
    console.warn
      <FilterInput> was created without expected prop 'filterText'



      at new FilterInput (src/components/FilterInput.svelte:132:12)
      at create_fragment (stories/FilterableList.svelte:56:16)
      at init (node_modules/svelte/internal/index.js:1465:37)
      at new FilterableList (stories/FilterableList.svelte:193:3)
      at getRenderedTree (node_modules/@storybook/addon-storyshots/dist/frameworks/svelte/renderTree.js:20:5)
      at node_modules/@storybook/addon-storyshots/dist/test-bodies.js:11:22
      at Object.<anonymous> (node_modules/@storybook/addon-storyshots/dist/api/snapshotsTestsTemplate.js:42:20)
```

It looks like maybe the story is lacking something? It would be good if we could make issues like that fail the CI, as they often indicate bugs.

wlach commented on src/components/FilterInput.svelte:
```diff
@@ -1,5 +1,5 @@
 <script>
-  import throttle from "just-throttle";
+  const throttle = require("just-throttle");
```
With lodash we can just do:

```suggestion
  import { throttle } from "lodash";
```

... and it'll work

wlach commented on src/components/Markdown.svelte:
```diff
@@ -1,5 +1,5 @@
 <script>
-  import marked from "marked";
+  const marked = require("marked");
```
```suggestion
  import { parseInline } from "marked";
```

DawoodShahat commented:
Here is what I updated upon @wlach review:

- Removed `just-throttle` dependency and instead used `_throttle` from `lodash`
- Fixed the warning that was generated by `FilterInput.svelte` component
- Used ES6 imports instead of CJS require

The CI will now fail when there is an unintended change in the markup. I think we should update the README and add some instructions on how to update the generated snapshots with your intended changes e.g. `npm run test:jest -- -u`.

Also one more thing, What do you think about deploying Stories to netlify? 

By the way thanks for the review, I spent a good amount of time on the error.

wlach commented:
> The CI will now fail when there is an unintended change in the markup. I think we should update the README and add some instructions on how to update the generated snapshots with your intended changes e.g. `npm run test:jest -- -u`.

Great idea, could you add something about this to the README? I think we can add them to the storybook: https://github.com/mozilla/glean-dictionary#storybook

> Also one more thing, What do you think about deploying Stories to netlify?

This is also a great idea! I think we could probably publish a static copy to `/storybook/`. I would obviously do this in a followup.

wlach reviewed (changes requested):
Code looks great, I think we just need an update to the README and we'll be good to land.

wlach commented on README.md:
````diff
@@ -49,6 +49,21 @@ run:
+npm run test:jest
+```
+
+If intended change is made to the markup, re-generate snapshots using the
+following command:
````
```suggestion
If you intentionally made a change to a component that results in a change 
to the output of the storybook snapshots, you can re-generate them using 
the following command:
```

wlach commented on README.md:
```diff
@@ -49,6 +49,21 @@ run:
 
+### Storybook Snapshot Testing
+
+To run snapshot tests, run:
```
```suggestion
To give us more confidence that changes don't unintentionally break the UI, we run [storybook snapshot tests](https://storybook.js.org/docs/react/workflows/snapshot-testing).

You can run them manually as follows:
```

wlach reviewed (approved):
Great stuff, just have some minor suggested changes to the README which I'll apply before landing.
